docs/explanation/holistic-vs-delta-charms.md

```diff
@@ -1,66 +1,230 @@
+
+Such applications should be profiled, and, based on the results, the charm developer may cache the Juju state, communicate less on the relation and [more out of band](https://github.com/charmed-hpc/slurm-charms/blob/edc47369560845fa54f81a2f02e0a3870b14302a/charms/slurmrestd/src/charm.py#L93-L98), or perhaps reconsider whether this key relation processing could be moved out of the reconciliation loop.
+Similar [performance concern](https://discourse.charmhub.io/t/performance-limitations-of-peer-relations-and-a-solution/18144) applies to peer relation processing for applications where hundreds of units may be expected.
+
+## Handling events individually
+
+A *delta-based* charm is when the charm handles each kind of Juju hook with a separate handler function, which does the minimum necessary to process that kind of event.
+
+Juju itself nudges charm authors in this direction.
+Juju provides specific event kinds that signal that one "thing" changed: `config-changed` says that a configuration value has changed, `relation-changed` says that relation data has changed, `pebble-ready` signals that the Pebble in the workload container has become ready, and so on.
 
-    def _update_charm(self, _: ops.EventBase):  # event parameter isn't used
-        redis_port = self.config.get('redis-port')
-        if not redis_port:
-            # pebble-ready happened first, wait for config-changed
-            return
+A delta charm may look like this:
 
-        # If both the Pebble container and config are ready, rewrite the
-        # container's config file and restart Redis if needed.
-        container = self.unit.get_container('redis')
-        try:
-            self._update_redis_config(container, redis_port)
-        except ops.pebble.ConnectionError:
-            # config-changed happened first, wait for pebble-ready
-            return
+### Example
+
+```py
+def __init__(self, framework: ops.Framework):
+    super().__init__(framework)
+    self.framework.observe(self.on.install, self._on_install)
+    self.framework.observe(self.on.start, self._on_start)
+
+    hostname = socket.getfqdn()
+    self.foo = FooRequires(self, "foo-relation", address=hostname)
+    self.framework.observe(self.foo.on.data_available, self._on_data_available)
+
+    self.bar = BarProvider(self, "bar-relation")
+    self.framework.observe(self.bar.on.create_bar, self._on_create_bar)
+
+def _on_install(self, event: ops.InstallEvent):
+    # Install the workload binary
+    apt.add_package("some-workload")
+
+def _on_start(self, event: ops.StartEvent):
+    # This unit has been started, peer relations are usable
+    if self.unit.is_leader():
+        ...
+
+def _on_data_available(self, event: DataAvailableEvent):
+    # Update the workload with event's data
+    self.workload.reconfigure(some_key=event.some_value)
+
+def _on_create_bar(self, event: CreateBarEvent):
+    # Provision a `Bar` resourse in the workload
+    self.workload.create_foo(event.some_field)
```

Comment:
Probably no "self" here, since there's no `self.workload = ` in `__init__`, and we're generally suggesting that the workload module has these sorts of things, without needing a class/object.

Reply:
I'll add `self.workload = ...` for parity with the holistic example.